### DISCUSSION TITLE
Hierarchical logging of computeds

### DISCUSSION THREAD

**Discussion by IanBellomy** [mobxjs/mobx#3467] [Q&A]
I have a complex network of derivations that runs great (thanks MobX!) but I'd like to log value changes with respect to the upstream values that caused the change, i.e. hierarchically / as a tree.

Example:
<img width="468" alt="CcZwT" src="https://user-images.githubusercontent.com/17239616/179318805-1d99cd30-3844-4f9d-8402-d43591242e03.png">

Before using Mobx, the computed functions were 'update' functions. I could construct an isomorphic log by relying on the call stack and pushing and popping contexts. (In most cases it was safe to assume that if an update was running there was a difference to log but in many cases a previous value was cached to compare against.)

e.g. 
````
updateA(){
    pushLog("A") // creates "A" log node.
    if ... updateX()
    else ... updateB();
    popLog()
}
updateB(){
    pushLog("B") // creates "B" log node.
    updateC()
    if ...  updateX()
    popLog(); // Remove the B log node, assuming everything else is pushing and popping as it should. 
}

updateC(){
   pushLog("C")
   //...
   popLog();
}

updateX(){
   pushLog("X")
   //...
   popLog();
}

````

With computeds and reactions though there is no clean call stack, so I'm not sure how to properly construct an isomorphic log tree.

I can create reactions for each computed but they will of course run at the end and without a guaranteed order and sans any kind of info on the chain-of-causation so to say. 

It smells like I might be able to _assemble_ the log tree after the fact if I kept track of the computed's debug names and then do some kind of dependency lookup in a reaction before appending the log entry — see if a previous entry in the log is a dependency, and if so, append the new log entry to that previous entry. But the log itself is an observable and updating it _after_ the transaction will create other complications — ideally the transaction that triggers the computed would contain the changes to the log. 

Are there any features in mobx that'd make this problem just go away or otherwise simplify it?

**Comment by mweststrate** [comment]
Did you check the `spy` API?

**Reply by IanBellomy** [reply]
Yes, but...

• The log will be user-facing (more on this below) — I understand spy is for dev/debugging? I'm assuming it has non trivial overhead.
• The goal is to log some computeds not all; though if the overhead is low, who cares of course.

Context:

Consider that in many boardgames many rules can be functionally expressed. e.g. In Catan, whether a player is a winner is a function of their total points and that total is a function of each players' connected roads (plus some other things). 

Presenting a UI-celebration when `player.won == true` is trivial but it would be _ideal_ to indicate that this winning was contingent on, for example, becoming the player with longest road and that was contingent on the player's road pieces. i.e.

> Placed road
>> Secured "Longest Road"! +10 points
>>> Victory!

To communicate these contingencies we need to know not just the initiating change (change in placed roads) but also the chain of derivations that led to a change in `player.won`. 

This is especially necessary where there exists multiple paths through the dependency tree from an initial change to an ending change. For example, imagine a Catan variant where placing a road could create a board situation where it was known to be logically impossible for any other player to win thereafter, resulting in a weird win-by-default...

> Placed road
>> All opponents boxed in.
>>> Victory!

**Reply by urugator** [reply]
Can't you return the log as part of the derived value?

```javascript
computed(() => {
  return {
    value: dep1.value + dep2.value
    log: [[dep1.log, dep2.log], 'This computed entry'],
  }
}, { equals: (a,b) => a.value === b.value })
```

**Reply by IanBellomy** [reply]
Using your example, it may be that a change in dep1 and/or dep2 is the reason for the computed update and we'd want log only the updated dependency that changed, not all of them. Perhaps I'm missing something? 

Just including dependency information in the computed result _is_ a lot cleaner than computeds having side effects though!

**Reply by urugator** [reply]
> dep1 and/or dep2 is the reason

Both of the deps can be changed in a single action, therefore both could be an equally good reason for it to run, therefore both of their logs are relevant. Otherwise you would need some additional heuristics like you want the first one that invalidated computed.
Keep in mind that computeds are lazy, they don't re-run immediately on depedency change, they just sit there in dirty state until something asks for the value (reaction) - that may not even happen if the computed is accessed conditionally.
The idea was that you set a new log together with the new value, so the unchanged dependencies would have an empty log. However I am not sure how you should go about clearing logs of the these "initial" observables (atoms).
I think you need some notion of a "turn" (a single derivation step), for which you get that log tree - I think that would map to a single batch in mobx (?), but I am not sure if that's what you really want/need. So in that case I quess you would clear all the logs at the start of every action?

**Reply by IanBellomy** [reply]
> I guess you would clear all the logs at the start of every action?

Bingo. Er... wait, are you talking about the return value of computed or the general 'log' I'm trying to build. 

> I think you need some notion of a "turn" (a single derivation step), for which you get that log tree - I think that would map to a single batch in mobx (?)

Bingo. MobX's transactions map well to player actions.

> The idea was that you set a new log together with the new value, so the unchanged dependencies would have an empty log.

Not sure I follow. In your example I'd assume that  `dep1.log` and`dep2.log` would either be a new value or a mobx-cached value such that `log: [[dep1.log, dep2.log], ...` would always have two pieces of data no matter what, making it impossible to know which (if not both) had changed... 🤔

> Keep in mind that computeds are lazy

Apologies, I'm taking for granted that these computeds are consumed by observer react components.

> Otherwise you would need some additional heuristics like you want the first one that invalidated computed.

Bingo.

> Both of the deps can be changed in a single action, therefore both could be an equally good reason for it to run, therefore both of their logs are relevant.

If I'm mistaken and it is the case in your example that only a changed dep will produce a log value, than yes I follow that it makes sense to say `log: [[dep1.log, dep2.log]...`, but it seems to me they will both always have values, making it impossible to distinguish between cases where only dep1 was responsible for a change or vice versa.

**Reply by urugator** [reply]
It's not like I think it through completely, just throwing ideas around...

```javascript
computed(() => {
  return {
    value: dep1.value + dep2.value
    log: dep1.log.length === 0 && dep.log.length === 0
      ? []
      : [[dep1.log, dep2.log], 'This computed entry'],
  }
}, { equals: comparator.struct }) // the log is also relevant

doSomethingAction() {
  // clear logs for all values, also invalidating all the computeds
  this.a = { value: this.a.value, logs: [] };
  this.b = { value: this.b.value, logs: [] };
  this.c = { value: this.c.value, logs: [] };
  // do something
  this.a = { value: "new", logs: ["new value set"] }
}
```

**Reply by IanBellomy** [reply]
Ah, I think I follow. It's an interesting idea I think and one that could solve the need [**edit**: maybe not] 🤔...  
 
Maybe because MobX provides `getDependencyTree(observable, prop)` and it provides the root cause in other places, I was hoping for a way to keep computeds focused on the values they need to provide and access the metadata about the overall procedures separately. 

Maybe it's just an aesthetic thing on this end / feels a little sad to have to invalidate everything and also a little weird to mix values with metadata that's of a kind available through other Mobx mechanisms 🤔... 

Really appreciate your thoughts though!

**Reply by IanBellomy** [reply]
(It may also feel smelly to me because I previously basically got the needed information for free using a call stack of update functions but now there's hoops for the simple things so that the hard things could be simple 🤔.)

**Reply by IanBellomy** [reply]
Looking closer I'm not sure how invalidation works.  🤔
• It is not possible to overwrite computed values  
• Modifying a `.log` property of a returned computed value won't have any effect on Mobx's cached values, and these internal cached values will be used next the computed is re-run, meaning the log value of the dependency will just be whatever it was / never empty.

**Comment by IanBellomy** [comment]
Question clarification and working (albeit sloppy) solution

Desired use case:

````typescript
// Within some action...

// set a property on an observable
observable.prop = 1

// get a tree/graph where 'prop' is the root and the specified computedProps, if changed, are the deepest nodes.
const results = causalTreeFor( 
   [observable, 'computedProp'],
   [observable, 'computedProp2'],
   [observable, 'computedProp3']
) 

// place the results where needed

````

I have a working solution but it's kinda of sloppy and involves looking at parts of computed Atoms that aren't public.
 
````typescript

/**
* Taking for granted that only one observable property changed, we should be able to trace a path back to that single prop.
* If there are alternate routes, it's ok we end up with one.
* I'm almost certain that the observing_ array order corresponds to the order of use in the computed, which is a nice bonus — it should mean the priority for items that dirty the computed (if multiple) ends up being their use in the computed. 
*/
function causalChain(atom:ComputedValue<any>):ComputedValue<any>[]{
    if(!atom || !atom.dependenciesState_) return [];
    const chain = [atom]
    let proximateCause = atom.observing_.find((a:ComputedValue<any>)=> !!a.dependenciesState_) as ComputedValue<any>
    while(proximateCause){
        chain.push(proximateCause)
        proximateCause = proximateCause.observing_.find((a:ComputedValue<any>)=> !!a.dependenciesState_) as ComputedValue<any>
    }
    return chain.reverse();
}

function causalNodeFromAtom(
    atom:ComputedValue<any>,
    listener:CausalNodeListener = (a,b)=>{return undefined}, // this can be used later to populate the node's description.
    salient?:boolean
){
    const prop = atom.name_.split(".")[1]
    return {
        id:atom.name_,
        atom,
        listener,
        prop,
        //@ts-ignore
        oldValue : atom.value_,
        newValue : undefined,
        //@ts-ignore
        update : ()=> atom.scope_?.[prop], //   :|
        //@ts-ignore
        children : new Set(),
        salient:salient ?? false,
        description: "" as undefined|string
    }
}
function causalTreeFor(...derivedValues:(ReturnType<typeof propListener>|false|undefined)[]){
    const causalNodes = derivedValues.filter(clean).map(o=> {
        return causalNodeFromAtom(
            getAtom(o[0],o[1]) as ComputedValue<any>, // This isn't necessarily true, but non-computed will ultimate be ignored/filtered out later.
            o[2],
            true
        )
    });

    // filter out nodes that we know for sure didn't change
    const dirtyNodes = causalNodes.filter((node)=> !!node.atom.dependenciesState_ );

    const nodes = new Map<string,CausalNode>(
        dirtyNodes.map(node=>[node.id,node])
    )

    const causalChains = dirtyNodes.map(node=>node.atom).map(causalChain)

    const causalChainNodes = causalChains.map(chain=>{
        return chain.map((a:ComputedValue<any>)=>{
            const node = nodes.get(a.name_) ?? causalNodeFromAtom(a)
            nodes.set(a.name_,node);
            return node;
        })
    })

    // update new values
    causalChainNodes.forEach(chain=>{
        chain.forEach(node=>{
            node.newValue = node.update();
            node.description = node.listener(node.oldValue,node.newValue)
        })
    })
    causalChainNodes.forEach(chain=>{
        chain.reduce(
            (previous:CausalNode|null,current)=>{
                previous?.children.add(current)
                return current
            },null)
    })

    const roots = Array.from(new Set(causalChainNodes.map(chain=>chain[0])))

    return roots

}
````

🤢🙃
